http-webhook.md

```diff
@@ -0,0 +1,339 @@
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+### 1.2. Relation to HTTP
+
+This specification prescribes rules constraining the use and handling of
+specific [HTTP methods][RFC7231-Section-4] and headers.
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection][#4-abuse-protection],
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
+the client MUST NOT follow any such redirection.
+
+If a delivery target has been retired, but the HTTP site still exists, the
+site SHOULD return a [410 "Gone"][410] status code and the sender SHOULD refrain
+from sending any further notifications.
+
+If the delivery target is unable to process the request due to exceeding a
+request rate limit, it SHOULD return a [429 Too Many Requests][429] status code
+and MUST include the [`Retry-After`][Retry-After] header. The sender MUST
+observe the value of the Retry-After header and refrain from sending further
+requests until the indicated time.
+
+If the delivery has been accepted and processed, and if the response carries a
+payload with processing details, the response MUST have the [200 OK][200] or
+[201 Created][201] status code. In this case, the response MUST carry a
+[`Content-Type`][Content-Type] header.
+
+If the delivery has been accepted and processed, but carries no payload, the
+response MUST have the [201 Created][201] or [204 No Content][204] status code.
+
+If the delivery has been accepted, but has not yet been processed or if the
+processing status is unknown, the response MUST have the [202 Accepted][202]
+status code.
```

Comment:
After submitting, I saw that the PR description says the spec is for "one-way event delivery".

I didn't find such a claim in the spec itself (unless you interpret "notification" as only applicable to past events). I understand how focusing on one-way delivery makes sense in the context of CloudEvents, but - if this spec is to be used for webhooks in general - I think you should consider the case of the delivery target sending a proper response that the sender uses to change its behaviour.

FWIW, the wikipedia definition of a webhook includes it as well: "A webhook [...] is a method of augmenting or **altering** the behaviour of  a web page, or web application [...]."

Reply:
@cneijenhuis If you can get a proper response, that's arguably just another HTTP request. RFC7231 covers that exhaustively. The point if this spec is to constrain RFC7231 to a profile that everyone agrees to for one-way event delivery, i.e. callbacks. 

Here's the place where we all make the spec and then I'm glad if someone summarizes that on Wikipedia - not the other way around :)

Reply:
Yes @cneijenhuis, it is not uncommon to have synchronous responses. Auth0 rules, for example, are synchronous. At the time of login, custom logic runs in rules (which are webhooks) which could prevent the login. We have many other synchronous use cases.

Reply:
@cneijenhuis @glennblock I understand that synchronous responses are being used, but CloudEvents defines a unidirectional and routable event flow (on top of which you can model async responses). The examples you cite are extensibility points where the web hook call performs a command execution in context. Those are not an events; events are facts, commands are intents.

Mind the headline here: "HTTP 1.1 Web Hooks for Event Delivery"

There might quite well be a derivative or parallel spec "HTTP 1.1 Web Hooks for Extensiblity Commands", but I would like to keep this one scoped to the one-way case that we keep symmetric to what we can easily achieve in AMQP, MQTT, NATS, etc.

Reply:
> I would like to keep this one scoped to the one-way case that we keep symmetric to what we can easily achieve in AMQP, MQTT, NATS, etc.

While I think it wouldn't be too hard to include the two-way cases, I fully understand that you want to keep the scope as small as possible.

If you allow me to make a final request, though, it would be to have the scope more explicitly in the title, e.g. "Web Hooks for Asynchronous Event Delivery", or "Web Hooks for One-Way Event Delivery". Let me try to explain:

> The examples you cite are extensibility points where the web hook call performs a command execution in context. Those are not an events; events are facts, commands are intents.

I have to disagree. All three examples actually send events, or facts, via the webhook.

* Twilio: Caller with ID foobar has called
* Auth0: User foobar has made a login attempt
* Dialogflow: User foobar has said: "Book a flight from LA to London"

These are arguably facts - events that have happened. If the use cases wouldn't require an immediate response, it would also be fine to implement them as asynchronous webhooks.

E.g. for a chatbot, both options work well:
* Chatbot receives event via asynchronous webhook. Chatbot responds by calling API.
* Chatbot receives event via synchronous webhook. Chatbot responds directly.

The event data would be exactly the same in both cases! Therefore, I think it is worth pointing out in the title that the spec is scoped to asynchronous/one-way event delivery.